### ISSUE TITLE
ONS-Trade-in-goods

### ISSUE THREAD

**Issue by ajtucker** [GSS-Cogs/family-trade#9]

**Comment by ajtucker** [comment]
Jenkins job result: FAILURE
Stage: Upload Tidy Data
[View full output](https://ci.floop.org.uk/job/GSS_data/job/Trade/job/ONS-UK-trade-in-services/86/)

Blocks #17

**Comment by ajtucker** [comment]
From the source spreadsheet:

> Data are non seasonally adjusted, current price in £'s.
> Note that the commodity are in £'s rounded to the nearest £ million.

- [ ] Change unit of measure to gbp (not gbp-million)

**Comment by ajtucker** [comment]
Jenkins job result: FAILURE
Stage: Upload Tidy Data
[View full output](https://ci.floop.org.uk/job/GSS_data/job/Trade/job/ONS_ABS/116/)

Blocks #32

**Comment by LPerryman** [comment]
Currently pipeline splits dataset into chunks and only outputs one trig and json file. changed pipeline so it now references the trade reference data but it did not like having several datasets but only 1 set of reference files. Changed so only 2 datasets are output  (Import and Export) and 2 sets of reference files. Run on Jenkins but got stuck on the validate CSV section as files are so large (over 7 million rows per dataset), this caused Jenkins to crash! i have put piepline on hold for now, see message from alex below.

**Comment by LPerryman** [comment]
Message from Alex pasted from Slack:
Right, for this specific dataset, ONS Trade in goods: country-by-commodity imports and exports, the only time we got this through the pipeline and into PMD was on 10th Jan 2019, based on a 10th Dec 2018 publication. The dataset was a new thing back then, but has been published monthly since.
At the time it caused us some grief because:
It takes quite a bit of memory to run the transform pipeline. I managed to reduce the time it took and get it working reasonably, but it can still cause problems when running other pipelines concurrently on Jenkins.
It blows up the initial CSV validation part of the pipeline (csvlint) as it’s so large, so we just ran it on a subset of the data as a minimal check. We can switch this off if needs be, or again look into better tools.
Uploading the resulting CSV is fine, though we had some problems with locking and concurrent builds at the time.
The conversion to RDF on PMD results in large temporary files (tens of GB) and took a few goes of multiple hours to get working.
The publication of the resulting massive data cube caused problems as it took so much longer than any other dataset and timed-out. In the end we manually published it, maybe using some server side db admin tools.
We did all the above partly to stress test how big a dataset we could cope with. The results were good, in that this really is a large dataset and we can still show it and query over it. It also lead to a feature request for the next version of PMD: publishing time should be constant, rather than depending on the size of the dataset;
The upshot:
It’d be good to be able to have the latest version published in PMD4 by end of the month, but not at the detriment of working on other trade datasets, both old and new.
We need to check with Swirrl how they plan to cope with getting this dataset into PMD4. They could copy the existing version, but that would mean that we have to update the existing version too; or they could take the CSV and upload to PMD4 themselves; or we could wait a bit for when we update our pipelines to upload to PMD4 directly.

**Comment by ajtucker** [comment]
Jenkins job result: FAILURE
Stage: Validate CSV
[View full output](https://ci.floop.org.uk/job/GSS_data/job/Trade/job/ONS-UK-trade-in-services/87/)

Blocks #17

**Comment by ajtucker** [comment]
Jenkins job result: FAILURE
Stage: Validate CSV
[View full output](https://ci.floop.org.uk/job/GSS_data/job/Trade/job/ONS-UK-trade-in-services/88/)

Blocks #17

**Comment by ajtucker** [comment]
Jenkins job result: FAILURE
Stage: Upload Tidy Data
[View full output](https://ci.floop.org.uk/job/GSS_data/job/Trade/job/ONS-UK-trade-in-services/89/)

Blocks #17

**Comment by ajtucker** [comment]
Jenkins job result: FAILURE
Stage: Test draft dataset
[View full output](https://ci.floop.org.uk/job/GSS_data/job/Trade/job/ONS-UK-trade-in-services/90/)

Blocks #17

**Comment by ajtucker** [comment]
Build was fixed!

**Comment by ajtucker** [comment]
There's some Jenkins job that has its wires crossed and is commenting on issues that don't belong to it. Re-opening as we need to:

- [ ] Update info.json to mark that we need to turn into RDF locally.
- [ ] Update info.json to mark that we don't want to upload to PMD for now. 
- [ ] Work on pmd-jenkins-pipeline to implement local conversion to (gzipped) RDF if marked as above. GSS-Cogs/pmd-jenkins-library#9
- [ ] Work on pmd-jenkins-pipeline to avoid uploading to PMD if marked as above. GSS-Cogs/pmd-jenkins-library#10

**Comment by IeuanMan** [comment]
@Tracey-B 

https://docs.google.com/document/d/1rjNAOXYPi08vAqjRB8EQXcYvsJlJ0H8dfI-xQS6vHfA/edit#heading=h.7783s8wimep8

BA Quality Assurance High Level Checks 
Dataset: Trade in goods: country-by-commodity exports
Completed by: Tracey Burgess
Date: 06/04/2020
PMD V4
Is the dataset listed  in PMD v4 staging?
Yes*
Does the dataset open?
Yes
Is there descriptive metadata on the PMD v4 staging landing page?
Yes
Does transformed info seem to match the original?
Yes*
Contents Modified Date
10/01/2019
PMD V3
Is the dataset listed  in PMD V3?
Yes 
Does the dataset open?
Yes
Is there descriptive metadata on the PMD v3 landing page?
Yes
Modified Date
10/01/2019
General
Is the dataset title meaningful?
Yes
Do the filters work & look sensible?
No**
Does the structure look sensible?
Yes

Overview
General comments on the dataset

*the data set is an amalgamation of 2 separate data sets (imports and exports) Title in PMD is Trade in goods: country-by-commodity imports and  exports.
 
This is not the most recent publication, both publications were published 11/03/2020, next release is due 09/04/2020.
Metadata
Is the Title, Publisher, Contact, Date Issued, description consistent with the source data

Better contact details are available on the landing page and XLS spreadsheet 
Dimensions
Do the column titles look sensible? Do the items look sensible? Does it match the original data?

**Not able to filter Product or Reference period 
**Do not need to filter by Measure type or Seasonal Adjustment 
**Is unit of measure and measure type required?
Observations
Is the PMD data consistent with the source data? Number of significant figures appropriate? Do they look sensible?

**Comment by RedWalters** [comment]
- [x] **Does the dataset show up or is there an error?**

Yes
PMD4 : http://cogs-staging-web.swirrlstaging.com/dataset/cube?uri=http%3A%2F%2Fgss-data.org.uk%2Fdata%2Fgss_data%2Ftrade%2Fons-trade-in-goods-catalog-entry

PMD3: http://gss-data.org.uk/dataset?uri=http%3A%2F%2Fgss-data.org.uk%2Fdata%2Fgss_data%2Ftrade%2Fons-trade-in-goods


- [x] **Is it in the list of datasets from the main page?**
- [x] **Is there descriptive metadata on the search page?**

- [x] **Does transformed info match the original?**

Yes - source data is in basic format so little tranformation was requried. No changes to content data.
'Direction' column removed as flow direction can be gleamed by the source file and added back later.

- [] **Does the tidydata download show all the data?**
No.

PMD4: Download functions but only downloads 14.8~ million lines as opposed to the 15.6~ million observations there should be.

PMD3: Download doesn't function ('SORRY' server error 500), likely due to timeout caused by the size of the dataset.

- [x] **Note differences in column/row headers**
Commodity and Country have been changed to CORD SITC and ONS Partner Geography respectively. No issues here. 
As mentioned 'Direction' column removed and Flow Direction added at the end.

- [x] **Are there multiple cubes? List titles of cubes**
N/A : One cube. 

- [x] **What needs differentiating? Totals**
N/A

- [x] **Are there titles that need harmonising? World/Worldwide**
CORD SITC is unclear without further information

- [x] **Does the structure look sensible?**
Yes. 

- [x] **Does the hierarchy work?**
Yes. 

- [x] **What needs further investigation or context?**
CORD SITC could either use a new column header or requires more readily available information.

- [x] **Any duplications?**
No.

- [x] **List any detailed metadata/methodology to add retrospectively**
Data is currently being stored as a single XLS inside of a zip file which seems unnecessary, will require changing file type being looked for if they decide to change this back to regular xls being hosted.


- [x] **Short metadata added**
N/A

**Comments:**

Only issues with this dataset seem to be with PMD and are most likely due to the size of the dataset.
Currently column sorting doesn't function in PMD4, again due to the size of the dataset.
Filters now work correctly

**Comment by JasonHowell** [comment]
28/04 - BA have confirmed good to go. One observation the landing page has it as 9/4, on PMD it is showing as 8/4.

**Comment by JasonHowell** [comment]
Closing issue as now been deployed onto PMDv4.

**Comment by osamede20** [comment]
Rework done on scripts
exports.py and imports.py called up on main.py
Cube class added

Legacy code removed

**Comment by rossbowen** [comment]
@AndrewtGSScogs 

- Where you specify `parent` also specify `codelist: false`.
- Where you specify `parent` you need to give the URI of a dimension, something like `http://gss-data.org.uk/def/trade/property/dimension/xxxxxxxxx`. Atm they point to `concept-scheme`s which are the URIs of the codelists.

**Comment by AndrewtGSScogs** [comment]
Updated info.json as per comments this is now ready for transform

**Comment by osamede20** [comment]
@AndrewtGSScogs 

Please I need some clarifications with regards the spec
1. #Create flow column: 
I thought there is an existing flow column which was was re-named flow direction?

2. #code list for commodity: 
Do I need to get code list from somewhere or keep the one or two or three letters (or figures) codes on the source data?

3. #creation of value column to sum export sums by year:
Please could clarify this further. Did you mean I should create a new column that sums up the exports from January to December for each respective year, say from 1997(Jan - Dec) and up to 2020(Jan - Dec)?

As I am combining both import and export, I guess this summation applies to both import and export sheets (or tabs)?

Thank you.

**Comment by AndrewtGSScogs** [comment]
Morning, @osamede20 

1. #apologies i believe the flow direction will be fine in this case.
2. #the commodity codelist has already been created so i can just point the info.json to it once the initial transform is complete. 
3. #Ideally if this could follow the same style as the trade-in-goods-by-industry-country-and-commodity so that we are being consistent across the dataset family. 

Hope this helps please let me know if not happy to discuss more.

**Comment by osamede20** [comment]
@AndrewtGSScogs 
Thanks for the clarification.

Transform has been re-written to meet the requirements of the provided spec for Stage 2 transform. Please let me know if changes are required.

**Comment by AndrewtGSScogs** [comment]
Dataset has passed the jenkins build and is displaying correctly. It just needs a final review before publishing 

Cheers,
A

**Comment by AndrewtGSScogs** [comment]
Althought this has successfully ran last week I needed to remove a suppression marker. Currently the latest attempt resulted in a out of memory issue i shall re run again this evening

**Comment by AndrewtGSScogs** [comment]
This is still an issue https://ci.floop.org.uk/job/GSS_data/job/Trade/job/ONS-Trade-in-goods/ i am getting a memory timeout due there being not enough space on jenkins. Not sure what the solution is to this

Thanks,
A

**Comment by AndrewtGSScogs** [comment]
Current solution is to zip up the folders first attempt has failed as it only looks for .csv files see Jenkins. Mike and Alex are looking in to alternative 

Cheers,
A

**Comment by ajtucker** [comment]
The latest run shows an out-of-memory error in the JVM.

I believe that's caused by the move from the stream based tools we were using (Apache Jena's RIOT) to a more closely coupled in-memory approach to parsing and serializing RDF, see https://github.com/GSS-Cogs/gss-jvm-build-tools/blob/main/src/main/scala/RdfRepo.scala.

One way to work around this blocker would be to split the output into multiple CSV files.

**Comment by LPerryman** [comment]
Now on PMD4 but have had to remove quite a few years worth of data as it was timing out when trying to publish.
Now also on the Beta site

**Comment by canwaf** [comment]
We should get all the data onto PMD, not just a subset. @robons, maybe the multi-file approach failed here and we need to go full accretive?

**Comment by robons** [comment]
@AndrewtGSScogs did you guys try the multi-graph approach to uploading the data to avoid timing out?

**Comment by ajtucker** [comment]
This used to work fine. I assume it's failing because were now trying to load the complete RDF cube into main memory only to retrieve the dataset and metadata graphs? See GSS-Cogs/pmd-jenkins-library#77

**Comment by robons** [comment]
> This used to work fine. I assume it's failing because were now trying to load the complete RDF cube into main memory only to retrieve the dataset and metadata graphs? See [GSS-Cogs/pmd-jenkins-library#77](https://github.com/GSS-Cogs/pmd-jenkins-library/issues/77)

That's kind of the reason it's failing, it's more that we've started running RDF queries on the whole of the ttl output which pretty much requires loading all the data into memory. But we have the multi-graph upload functionality to help fix that problem.

**Comment by ajtucker** [comment]
> > This used to work fine. I assume it's failing because were now trying to load the complete RDF cube into main memory only to retrieve the dataset and metadata graphs? See [GSS-Cogs/pmd-jenkins-library#77](https://github.com/GSS-Cogs/pmd-jenkins-library/issues/77)
> 
> That's kind of the reason it's failing, it's more that we've started running RDF queries on the whole of the ttl output which pretty much requires loading all the data into memory. But we have the multi-graph upload functionality to help fix that problem.

Pretty sure I said that a month or so back? https://github.com/GSS-Cogs/family-trade/issues/9#issuecomment-831115042

**Comment by JasonHowell** [comment]
Closing as it's on Beta env.
